Unify tools building

ishitatsuyuki opened the pull request:
Close #41601 

Time saving for up to 10 minutes. Cargo is now only compiled once.

Downsides:

- Out of tree Cargo.lock maintenance
- Cargo.toml `[replace]` version maintenance

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

ishitatsuyuki commented:
r? @alexcrichton 

Currently the RLS dependencies are fucked up, and I have opened rust-lang-nursery/rls#285 to adapt the current state. I'm too lazy to hand-rewrite `Cargo.lock`, so this is blocked on that PR.

kennytm commented on src/bootstrap/compile.rs:
```diff
@@ -60,13 +60,13 @@ pub fn std(build: &Build, target: &str, compiler: &Compiler) {
+    // so we don't set this variable during stage0 where llvm-config is
+    // missing
+    // We also only build the runtimes when --enable-sanitizers (or its
+    // config.toml equivalent) is used
```
What's with all these spacing changes and reformatting?

ishitatsuyuki replied:
I used rustfmt, but seems it screwed up something else. Will remove.

ishitatsuyuki commented:
According to the build results, this saves up to 300 seconds. Not sure if this worth the out-of-tree lockfile maintenance cost.

ishitatsuyuki commented:
r? @alexcrichton

Review points:
- Cargo.lock now diverges from upstream. It also needs to be updated on each submodule update. (A simple `cargo update` in the directory is enough)
- The `[replace]` section also needs to be updated on version bump. This shouldn't be a big hit though.

ishitatsuyuki commented:
By the way, this could be a little bit even faster by swapping the build order of rls and cargo. In that way, it should have more parallelism since rls includes the cargo build and more crates are built in one pass.

alexcrichton commented:
Thanks for the PR! I think this is the lowest hanging fruit of speeding up the dist builders right now and this is also [an issue that must be resolved](https://github.com/rust-lang/rust/issues/41198) before the RLS hits stable.

The downsides here are unfortunate, namely disregarding the lock files in the rls/cargo repositories. Unfortunately though I don't see another way of doing this. I discussed this strategy with @brson and @nrc yesterday and we're comfortable having a separate lock file in-tree for the Cargo/RLS binaries that we ship.

Could this actually, though, move Cargo/RLS into the global rustc workspace? That way all their dependencies will get vendored as usual and should be buildable from source tarballs as well.

alexcrichton commented:
The end goal is to have everything in one workspace, and I don't believe many issues will arise. Mind trying it out?

ishitatsuyuki commented:
Old branch backed up, let's see what will happen now.

ishitatsuyuki commented:
@alexcrichton the well known submodule delayed manifest problem is here, how should I deal with that?

ishitatsuyuki commented:
```
[01:09:29] error[E0523]: found two different crates with name `rustc_serialize` that are not distinguished by differing `-C metadata`. This will result in symbol conflicts between the two.
[01:09:29]   --> /checkout/src/tools/rls/src/build.rs:13:1
[01:09:29]    |
[01:09:29] 13 | extern crate rustc_driver;
[01:09:29]    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
[01:09:29] 
[01:09:29] error: Could not compile `rls`.
```

Something's wrong. No idea.

Mark-Simulacrum commented:
@alexcrichton Is there a chance you could help diagnose the failure above?

ishitatsuyuki commented:
@alexcrichton I recommend to roll back to the original implementation (backed up as a [branch](https://github.com/ishitatsuyuki/rust/tree/unify-tools-old)). Here are the points:

- `rustc_serialize` is complexly reused/re-exported inside the compiler code, and it caused the linkage problem.
- We shouldn't give stage2 tools the same workspace, since they're intended for compiling independently from rustc.
- Putting submodules into workspace causes bootstrap failure (missing submodule = missing Cargo.toml).
- Regarding your intense to vendor stage2 deps, I believe it can be just solved by running cargo-vendor twice.

alexcrichton commented:
It's worth being clear up front that rustbuild with respect to crates.io is pretty precarious right now, and builds like this are showing how it's really starting to stress many underlying assumptions. I'm basically just saying that this is not going to be an easy integration do "do right," and there's likely tons of little surprises along the way like this.

Unfortunately I don't know why that error is showing up. The only difference here is a change in workspace, but today we're already compiling rustc-serialize a bunch of times. For example the compiler currently links to rustc-serialize 0.3.23 and so does the rls, so why the linking error only shows up with this PR and not currently on master is confusing to me.

In general we want to make these all one workspace to have one location that we're managing dependencies, not a whole bunch. The linkage problem here sounds like either a bug in the compiler or a bug in the setup, I'm not sure which. These tools are intended to be compiled with rustc as well for a full distribution, so we want them in the same workspace. I had forgotten about the submodule problem, yes, but I think we'll just need to bite the bullet and move git submodule management into `bootstrap.py` from `src/bootstrap/lib.rs`. It's worth pointing out that `cargo-vendor` does not currently support being invoked twice, there's an open bug about that.

bors commented:
:umbrella: The latest upstream changes (presumably #41846) made this pull request unmergeable. Please resolve the merge conflicts.

ishitatsuyuki commented:
It seems rustc_private content are conflicting with the workspace ones. Will investigate.

Additional personal note: one stage2 leftover in tidy.rs.

bors commented:
:umbrella: The latest upstream changes (presumably #41900) made this pull request unmergeable. Please resolve the merge conflicts.

ishitatsuyuki commented:
Guess what...

```
ishitatsuyuki@X1Yoga  ~/D/rls   master  cargo build --release        
   Compiling rls v0.1.0 (file:///home/ishitatsuyuki/Documents/rls)
error[E0523]: found two different crates with name `bitflags` that are not distinguished by differing `-C metadata`. This will result in symbol conflicts between the two.
  --> src/build.rs:12:1
   |
12 | extern crate rustc;
   | ^^^^^^^^^^^^^^^^^^^

error: Could not compile `rls`.

To learn more, run the command again with --verbose.
```

**This was a rustc/cargo bug, or RLS's toml being wrong.**

(I set opt_level to 2 in RLS's Cargo.toml so it matches what nightly uses, and this was reproduced.)

ishitatsuyuki commented:
https://github.com/rust-lang/cargo/issues/4028. P-High if possible.

alexcrichton commented:
Thanks for tracking that down, definitely sounds like a bug!

I think our solution here, though, should probably not depend on Cargo itself. All we need to do is to change the `-C metadata` argument to the compiler. I think we can intercept `metadata=...`  in `src/bootstrap/bin/rustc.rs` when compiling libstd/libtest/librustc and change the argument to something different (specific to rustc itself)

ishitatsuyuki commented:
🎁BUILD PASSED.

The last thing on the checklist: submodule fetching and workspace issues. @alexcricton how would you like to tackle that? I would appreciate a detailed explanation.

alexcrichton commented:
@ishitatsuyuki I think we'll basically just have to move [this code](https://github.com/rust-lang/rust/blob/e19ccb71c8427135a69d874623af68422aeeb9e9/src/bootstrap/lib.rs#L296-L409) [here](https://github.com/rust-lang/rust/blob/e19ccb71c8427135a69d874623af68422aeeb9e9/src/bootstrap/bootstrap.py#L599), essentially just making git submodule operations the first thing that rustbuild does.

alexcrichton commented on src/bootstrap/check.rs:
```diff
@@ -106,7 +106,8 @@ pub fn cargo(build: &Build, stage: u32, host: &str) {
     let mut cargo = build.cargo(compiler, Mode::Tool, host, "test");
-    cargo.arg("--manifest-path").arg(build.src.join("src/tools/cargo/Cargo.toml"));
+    cargo.arg("--manifest-path").arg(build.src.join("src/Cargo.toml"));
+    cargo.arg("--project").arg("cargo");
```
I think in theory this shouldn't need to be modified, but did you find it did need to be?

ishitatsuyuki replied:
Not strictly necessary; but well, unifying the look like below.

alexcrichton replied:
Ok, can this be backed out then? I think this has clearer intent that Cargo's being built (pointing at Cargo's manifest)

alexcrichton commented on src/bootstrap/compile.rs:
```diff
@@ -462,7 +462,8 @@ pub fn tool(build: &Build, stage: u32, target: &str, tool: &str) {
     let dir = build.src.join("src/tools").join(tool);
-    cargo.arg("--manifest-path").arg(dir.join("Cargo.toml"));
+    cargo.arg("--manifest-path").arg(build.src.join("src/Cargo.toml"));
+    cargo.arg("--package").arg(tool);
```
(same comment as above)

ishitatsuyuki replied:
This is the key point of the PR: by making it workspace aware we can save the double rebuild.

ishitatsuyuki replied:
Turns out that Cargo tries to find the workspace toml and use it automatically if it is needed.

alexcrichton commented on src/bootstrap/compile.rs:
```diff
@@ -474,17 +475,19 @@ pub fn tool(build: &Build, stage: u32, target: &str, tool: &str) {
     }
 
-    cargo.env("CFG_RELEASE_CHANNEL", &build.config.channel);
+    if stage == 2 {
```
How come this needs to be gated by `stage == 2`?

ishitatsuyuki replied:
Because it's what the addition intended to be: adding version hash to stage2 tools. I have tested this and it just works without any problem.

alexcrichton replied:
Could we leave this as is? I don't believe there's any reason to omit this information in other stages?

alexcrichton commented on src/bootstrap/lib.rs:
```diff
@@ -482,6 +482,12 @@ impl Build {
 
+        // Since the metadata of dependencies can conflict, we're adding a
+        // special suffix to make it unique.
+        if mode == Mode::Librustc {
```
Could this actually be for `mode != Mode::Tool` (e.g. also apply to libstd/libtest)

ishitatsuyuki replied:
Yeah, I originally thought that libstd has absolutely zero dependencies, but turns out it's not. Fortunately, either the version is set to 0.0.0 or the feature set is different so the metadata differs. Anyway, I will enable it for other types too due to the potential risk of breaking it in future.

alexcrichton commented on src/bootstrap/lib.rs:
```diff
@@ -482,6 +482,12 @@ impl Build {
         }
 
+        // Since the metadata of dependencies can conflict, we're adding a
+        // special suffix to make it unique.
```
Could this be:

> Currently the compiler depends on crates from crates.io, and then other crates can depend on the compiler (e.g. `proc-macro` crates. Let's say, for example that rustc itself depends on the `bitflags` crate. If an external crate then depends on the `bitflags` crate as well, we need to make sure they don't conflict, even if they pick the same verison of `bitflags`. We'll want to make sure that e.g. a plugin and rustc each get their own copy of bitflags.
>
> Cargo ensures that this works in general through the `-C metadata` flag. This flag will frob the symbols in the binary to make sure they're different, even though the source code is the exact same. To solve this problem for the compiler we extend Cargo's already-passed `-C metadata` flag with our own. Our `rustc.rs` wrapper around the actual `rustc` will detect `-C metadata` being passed and frob it with this extra string we're passing in.

bors commented:
:umbrella: The latest upstream changes (presumably #41955) made this pull request unmergeable. Please resolve the merge conflicts.

ishitatsuyuki commented:
I have roughly tested this and the new submodule handling should be OK. Please comment for any other concerns.

bors commented:
:umbrella: The latest upstream changes (presumably #41847) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented on src/bootstrap/bootstrap.py:
```diff
@@ -529,6 +534,30 @@ def build_triple(self):
+        if os.path.isdir(os.path.join(self.rust_root, ".git")) and \
+            (self.get_toml('submodules') != "false" and \
+                self.get_mk('CFG_DISABLE_MANAGE_SUBMODULES') != "1"):
+            print('Updating submodules')
```
Could the indentation here be reduced by returning early with the conditions above?

alexcrichton commented on src/bootstrap/bootstrap.py:
```diff
@@ -529,6 +534,30 @@ def build_triple(self):
+            (self.get_toml('submodules') != "false" and \
+                self.get_mk('CFG_DISABLE_MANAGE_SUBMODULES') != "1"):
+            print('Updating submodules')
+            self.run(["git", "-C", self.rust_root, "submodule", "-q", "sync"])
```
The current implementation runs `git submodule status`, is there a reason that behavior wasn't ported over?

ishitatsuyuki replied:
... because there was a FIXME there claiming `git submodule status` is slow. I made this identical to that without the potential slowness.

alexcrichton replied:
Ok, we can see how this works out

alexcrichton commented on src/bootstrap/bin/rustc.rs:
```diff
@@ -38,7 +38,23 @@ use std::path::PathBuf;
+    // Insert the metadata suffix
+    if let Ok(s) = env::var("RUSTC_METADATA_SUFFIX") {
+        for i in 1..args.len() {
+            // Dirty code for borrowing issues
```
This comment doesn't seem too helpful? Could you instead reference the comment in `lib.rs`?

alexcrichton commented on src/bootstrap/lib.rs:
```diff
@@ -475,12 +358,30 @@ impl Build {
+            // already-passed -C metadata flag with our own. Our rustc.rs
+            // wrapper around the actual rustc will detect -C metadata being
+            // passed and frob it with this extra string we're passing in.
+            cargo.env("RUSTC_METADATA_SUFFIX", "rustc".to_string());
```
I believe you can elide the `to_string` here

bors commented:
:umbrella: The latest upstream changes (presumably #41987) made this pull request unmergeable. Please resolve the merge conflicts.

ishitatsuyuki commented:
r? @alexcrichton 

I have removed all the command invocation changes since it turns out that they're not necessary. I have kept each change as an commit for easy review, but please tell me when you're ready so I can rebase/squash.

ishitatsuyuki commented:
Hmm.. something must be wrong, this is taking longer than expected.

alexcrichton commented:
Looks like the build succeeded? I tested this out locally and found two things:

* I think tidy fails in `src/bootstrap/bootstrap.py` for a "line too long"
* This *almost* passes the license check of tidy, the `strings` crate being vendored doesn't have a license listed. On [master](https://github.com/nrc/strings.rs/blob/master/Cargo.toml) though it's listed as apache/mit, though. @nrc mind publishing a new version with the license listed in `Cargo.toml`? We can always add an exception, but it'd be best if we didn't have to.

ishitatsuyuki commented:
This doesn't save time at all for now due to rust-lang/cargo#4048. Not a blocker, so I'm going to just make this mergeable as-is.

The tidy is a quite hard problem since it causes more weird indentation if I broke anywhere else. I will try to do my best.

ishitatsuyuki commented:
Ready to merge, except for the license tidy issue.

alexcrichton commented:
@bors: r+

Ok, let's see how far this gets and see if it will fail before tidy, just to weed out anything else if it's still there. Thanks again for taking this on!

bors commented:
:pushpin: Commit ef3d8b0 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton ef3d8b0fc6fee5bacbfbcd59e208c08561b88452 -->

bors commented:
:hourglass: Testing commit ef3d8b0 with merge 1b20dfe...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3268)

alexcrichton commented:
@bors: retry

* network issues https://github.com/rust-lang/rust/issues/40474

bors commented:
:umbrella: The latest upstream changes (presumably #41843) made this pull request unmergeable. Please resolve the merge conflicts.

aidanhs commented on src/bootstrap/bootstrap.py:
```diff
@@ -297,8 +298,10 @@ def bin_root(self):
         for line in self.config_toml.splitlines():
-            if line.startswith(key + ' ='):
-                return self.get_string(line)
+            match = re.match(r'^{}\s*=(.*)$', line)
```
This looks like it's meant to be `.format(key)`?

aidanhs commented on src/bootstrap/bootstrap.py:
```diff
@@ -529,6 +534,32 @@ def build_triple(self):
         return "{}-{}".format(cputype, ostype)
 
+    def update_submodules(self):
+        if (not os.path.isdir(os.path.join(self.rust_root, ".git"))) or \
```
It's not valid to assume `.git` is a directory - if the rust repo itself is a git submodule and you're using particular versions of git, `.git` will be a file containing a single line (the path pointing to the *actual* `.git` directory).

ishitatsuyuki commented:
Thanks for pointing out! r- if needed, will correct it this night.

aidanhs commented:
I feel a bit cautious about the rework of how git submodules are handled, given it's happening in parallel with the handling code being migrated to a) a different stage in the build process and b) a different language - git tends to be subtle in unfortunate ways. I'd personally prefer to see a separate PR, or even a separation into two individual commits (one direct translation, a followup to rework) in this one, to give it appropriate scrutiny. Just a thought.

@ishitatsuyuki the merge conflicts will block it so no worries

ishitatsuyuki commented:
I have carefully separated the submodule changes to a dedicated commit, so it would be easy to revert only that one. Anyway, the CI handles submodules in a separate script, so this cannot be auto tested.

alexcrichton commented:
@aidanhs ah unfortunately the submodule changes can't happen later, they have to happen as part of this commit. What this is doing is placing everything into the same workspace (including submodules like the rls and Cargo). If we leave them in submodules then a fresh checkout will fail to build b/c to build the build system we need submodules, but to update submodules we'd need the build system :(

aidanhs commented:
@alexcrichton not sure if I'm understanding you. My proposal would be to move the submodule handling into `bootstrap.py` (as you say, that's necessary), but make it a(n almost) line-by-line translation into Python, i.e. don't change it to do new things with the `--get-regexp`, `-C` etc arguments.

alexcrichton commented:
Oh sorry I was just responding to "I'd personally prefer to see a separate PR" and how I think it may not be possible, seems fine to me though to stay line-for-line equivalent and then the logic can change in a future PR!

ishitatsuyuki commented:
Congratulations, #41843 hopefully just broke the submodule handling. This is now urgent :P

ishitatsuyuki commented:
I have no idea but adding the rust-installer module seems to make Cargo infinite-loop on dependency resolution. @alexcrichton can you dig into that?

alexcrichton commented:
@ishitatsuyuki are you looking into preserving the git logic line-for-line? If not I can try to help rebase/move this forward

alexcrichton commented:
To get past the tidy check it's ok to add an exception to the license check for now, it's MIT/Apache-2.0 anyway just needs to be updated.

ishitatsuyuki commented:
I have corrected the bootstrap issues and rebased **removing Cargo.lock**.

@alexcrichton:
 - cherry pick the bootstrap commit if you consider it urgent.
 - please checkout the current head and investigate Cargo deadlocks.

alexcrichton commented:
Turns out this is an unsatisfiable dependency graph, I've [opened an issue](https://github.com/rust-lang/cargo/issues/4066) to not look like we're going into an infinite loop here.

The problem is that `mdbook` right now requires `clap ~2.19.0` but rust-installer requires `clap 2.22.1`, and these are incompatible version constraints.

ishitatsuyuki commented:
I don't think so. mdbook should be using the default caret, and the current lockfile is happily resolving to clap 2.22.1.

I don't think there would be impossible dependency graph for such reasons, since it's allowed to have multiple versions of crate in one binary.

alexcrichton commented:
Oh oops read the log wrong, it's [racer](https://crates.io/crates/racer/2.0.6) that depends on `clap ~2.19.0`

alexcrichton commented:
I've [pushed a commit](https://github.com/rust-lang/rust-installer/commit/daa2a05ebe7b8d07a309e8891ebc548652362954) to rust-installer to relax the constraint on clap to allow 2.19, mind updating the submodule here? I think we'll be good to go with that (resolution finished and the build is proceeding locally)

ishitatsuyuki commented:
That sounds like something that should diverge to two isn't being so due to time complexity blowing up. I would probably open a PR to correct the racer dependencies being too strict too.

Rebase in progress, just a minute...

alexcrichton commented:
Yeah it'd be great if the racer dep could be relaxed, I just figured the quickest solution is to update rust-installer (tested locally to make sure it works)

And yes Cargo shouldn't take so long to say that this is an un-resolvable graph, that's https://github.com/rust-lang/cargo/issues/4066

ishitatsuyuki commented:
Ready to roll.

alexcrichton commented:
@ishitatsuyuki did you want to update the git logic to be line-for-line equivalent to the current manifestation in rustbuild and update the logic later in a future PR? (discussion [here](https://github.com/rust-lang/rust/pull/41639#issuecomment-301667519))

ishitatsuyuki commented:
I don't want to do line-by-line translation personally (due to I having not so much time these days). I'm aware that this probably still has some flaw not found in review, but it's working for most of my local usecases and I consider it OK.

alexcrichton commented:
Ok I think I agree with @aidanhs that we don't want to land this until it's the same as before. The submodule logic has been careful crafted over a long time, so without a reason to change it I don't think we should. If you're busy though I can push up the relevant logic.

ishitatsuyuki commented:
I understand. I will create a line2line version tomorrow.

alexcrichton commented:
@bors: r+

Ah no worries, I went ahead and did so

bors commented:
:pushpin: Commit db69d89 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton db69d894e73aa5ae2315f7e2490104a3e5503afb -->

alexcrichton commented:
@bors: p=1

(fixes breakage on master related to submodules)

bors commented:
:hourglass: Testing commit db69d89 with merge be1d1aa...

jwilm commented:
I just published an update for Racer relaxing dependencies. [2.0.7 on crates.io]

[2.0.7 on crates.io]: https://crates.io/crates/racer/2.0.7

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/233350437?utm_source=github_status&utm_medium=notification)

Mark-Simulacrum commented:
I suspect this is a spurious failure since we timed out, so retrying.

@bors retry

bors commented:
:hourglass: Testing commit db69d89 with merge 31bfa96...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3293), [status-travis](https://travis-ci.org/rust-lang/rust/builds/233450896?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 31bfa964b8e67805df759450e0aca1690721f91c to master...

alexcrichton commented:
🎊 

Thanks again for working on this @ishitatsuyuki!
